cdap-app-templates/cdap-etl/cdap-calcite/src/main/java/io/cdap/cdap/datapipeline/relational/CDAPCalciteSchema.java

```diff
@@ -0,0 +1,235 @@
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static RelDataType getDatatypeFromSchema(Schema cdapSchema, boolean isNullable)
+            throws SQLDialectException {
+        Schema.Type cdapDatatype = cdapSchema.getType();
+        RelDataType calciteDatatype;
+
+        switch (cdapDatatype) {
+            case NULL:
+                calciteDatatype = new BasicSqlType(RelDataTypeSystem.DEFAULT, SqlTypeName.NULL);
+                break;
+            case BOOLEAN:
+                calciteDatatype = new BasicSqlType(RelDataTypeSystem.DEFAULT, SqlTypeName.BOOLEAN);
+                break;
+            case INT:
+                calciteDatatype = getDatatypeFromInt(cdapSchema);
+                break;
+            case LONG:
+                calciteDatatype = getDatatypeFromLong(cdapSchema);
+                break;
+            case FLOAT:
+                calciteDatatype = new BasicSqlType(RelDataTypeSystem.DEFAULT, SqlTypeName.FLOAT);
+                break;
+            case DOUBLE:
+                calciteDatatype = new BasicSqlType(RelDataTypeSystem.DEFAULT, SqlTypeName.DOUBLE);
+                break;
+            case BYTES:
+                calciteDatatype = getDatatypeFromBytes(cdapSchema);
+                break;
+            case STRING:
+                calciteDatatype = getDatatypeFromString(cdapSchema);
+                break;
+            case ENUM:
+                // TODO: Add implementation
+                calciteDatatype = null;
+                break;
+            case ARRAY:
+                calciteDatatype = new ArraySqlType(getDatatypeFromSchema(cdapSchema.getComponentSchema()), isNullable);
+                break;
+            case MAP:
+                Schema cdapKeySchema = cdapSchema.getMapSchema().getKey();
+                Schema cdapValueSchema = cdapKeySchema.getMapSchema().getValue();
+                calciteDatatype = new MapSqlType(getDatatypeFromSchema(cdapKeySchema),
+                        getDatatypeFromSchema(cdapValueSchema), isNullable);
+                break;
+            case RECORD:
+                calciteDatatype = new RelRecordType(getListOfFields(cdapSchema.getFields()));
+                break;
+            case UNION:
+                if (cdapSchema.isNullable()) {
+                    calciteDatatype = getDatatypeFromSchema(cdapSchema.getNonNullable(), true);
+                } else {
+                    // TODO: Add implementation
+                    calciteDatatype = null;
+                }
+                break;
+            default:
+                throw new SQLDialectException("Invalid data type found in schema: " + cdapDatatype);
+        }
+
+        return calciteDatatype;
+    }
+
+    private static RelDataType getDatatypeFromInt(Schema cdapSchema) {
+        Schema.LogicalType logicalType = cdapSchema.getLogicalType();
+        RelDataType calciteDatatype;
+
+        if (logicalType != null) {
+            switch (logicalType) {
```

Comment:
I would do it reverse. First check the logical type, then physical. Otherwise if we decide to change which physical type to use for given logical, it would be really hard to debug